test/jdk/java/nio/channels/DatagramChannel/SelectWhenRefused.java

```diff
@@ -49,43 +55,62 @@ public static void main(String[] args) throws IOException {
             dc.configureBlocking(false);
             dc.register(sel, SelectionKey.OP_READ);
 
-            /* Test 1: not connected so ICMP port unreachable should not be received */
-            sendDatagram(dc, refuser);
-            int n = sel.select(2000);
-            if (n > 0) {
-                sel.selectedKeys().clear();
-                // BindException will be thrown if another service is using
-                // our expected refuser port, cannot run just exit.
-                DatagramChannel.open().bind(refuser).close();
-                throw new RuntimeException("Unexpected wakeup");
-            }
+            for (;;) {
+                /* Test 1: not connected so ICMP port unreachable should not be received */
+                sendDatagram(dc, refuser);
+                int n = sel.select(2000);
+                if (n > 0) {
+                    ignoreStrayWakeup = checkUnexpectedWakeup(dc, sel.selectedKeys());
+                    sel.selectedKeys().clear();
+                    if (ignoreStrayWakeup) {
+                        continue;
+                    }
+                    // BindException will be thrown if another service is using
+                    // our expected refuser port, cannot run just exit.
+                    DatagramChannel.open().bind(refuser).close();
+                    throw new RuntimeException("Unexpected wakeup");
+                }
 
-            /* Test 2: connected so ICMP port unreachable may be received */
-            dc.connect(refuser);
-            try {
+                /* Test 2: connected so ICMP port unreachable may be received */
+                dc.connect(refuser);
+                try {
+                    sendDatagram(dc, refuser);
+                    n = sel.select(2000);
+                    if (n > 0) {
+                        sel.selectedKeys().clear();
+                        try {
+                            ByteBuffer buf = ByteBuffer.allocate(100);
+                            SocketAddress sa = dc.receive(buf);
```

Comment:
This changes the test to use receive instead of read. I am not sure we want to do that. It's true that if we use read we won't be able to tell where the message comes from - but we're connected, so in theory it could only come from the refuser address, which would indicate that some other process has reused the port. So it should be enough to call read() and print the content of the received buffer. @AlanBateman can we use receive() here instead of read() without changing the nature of the test?